### PR TITLE
Feat/withdraw from private key

### LINKED ISSUE #29: Withdraw from another address not working (+other issues)
Precoditions:
* User already has burner wallet with some pDAI/sunDAI
* QR-code with private key for another wallet with some funds

Reproduction:
1. Scan QR-code from prerequisites
1. Withdraw screen opens - OK
1. Try to withdraw funds to your wallet
1. No mater what amount user wants to withdraw error is shown - WRONG - user should be able to withdraw funds

See attached screenshot with highlighted issues:
1. `xdai` instead of `pDai` or `sunDAI`
1. It shows "target" (user's) balance but that's not clear
1. Doesn't show balance of "source" account (the one from QR code)
1. Error message is shown no mater the amount user enters

Screenshot_20190413-160428_mod

### PR DESCRIPTION
**Pull request by TimDaub**
This PR allows to withdraw pDAI from a private key. It fixes #29. It is based on #35.

TODOs:
- [x] Investigate whether xdai.io can also withdraw DAI and ETH from private key
- [x] Remove TODOs from the code
- [x] Test on browser without metamask
- [x] Make commit history pretty
- [x] Get leap-core#89 merged and published. Pin new leap-core version in package.json if PR goes through.
- [x] Use: https://github.com/leapdao/leap-core/blob/master/lib/transaction.js#L92
- [x] Pin leap-core@0.29.0

## Does the burner-wallet allow to withdraw DAI or ETH from a private key?

I conducted a very simple test where I tried to scan a private key on xdai.io and it in fact only gave me the option to withdraw xDai. So no, it doesn't.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -2024,80 +2026,65 @@ render() {
...
-    });
+  if (metaAccount) {
+    await tokenSendV2(account, to, value, 0, xdaiweb3, web3, metaAccount.privateKey)
+  } else {
+    await tokenSendV2(account, to, value, 0, xdaiweb3, web3, null)
+  }
+  // NOTE: The callback cb is not used correctly in the format 
+  // cb(error, receipt) throughout the app. We hence cannot send errors in
+  // the callback :(
+  // TODO: leap-core doesn't return receipts as part of their architecture.
```
Comment (TimDaub) [inline src/App.js]: I'll address this in another PR.

File: src/App.js [hunk 2]
```diff
@@ -2025,80 +2026,65 @@ render() {
...
-    ).then(receipt => {
-      cb(receipt);
-    });
+  let receipt;
+  if (metaAccount) {
+    receipt = await tokenSendV2(account, to, value, 0, xdaiweb3, web3, metaAccount.privateKey)
+  } else {
+    receipt = await tokenSendV2(account, to, value, 0, xdaiweb3, web3, null)
+  }
+  // NOTE: The callback cb is not used correctly in the format
```
Comment (sunify) [inline src/App.js]:
```es6
const receipt = await tokenSendV2(account, to, value, 0, xdaiweb3, web3, metaAccount && metaAccount.privateKey)
```

Much simpler

File: src/App.js [hunk 3]
```diff
@@ -2025,80 +2026,65 @@ render() {
...
+
+  let signedTx;
+  if (!privateKey) {
+    signedTx = await transaction.signWeb3(web3)
+  } else {
+    const privs = [];
+    for(const input of transaction.inputs){
+      privs.push(privateKey);
+    }
+    signedTx = await transaction.sign(privs)
```
Comment (sunify) [inline src/App.js]: There is a method `singAll` for this (`await transaction.signAll(privateKey)`. So you can be written like this:

```es6
const signedTx = privateKey ? await transaction.signWeb3(web3) : await transaction.signAll(privateKey);
```

File: src/components/WithdrawFromPrivate.js [hunk 4]
```diff
@@ -65,50 +62,33 @@ export default class SendToAddress extends React.Component {
...
-                  message: 'Withdrawn! '+receipt.transactionHash,
-                });
-              }
-
-            })
+        const weiAmount = web3.utils.toWei(amount, "ether")
+        await tokenSendV2(
+          fromAddress,
+          address,
+          weiAmount,
```
Comment (sunify) [inline src/components/WithdrawFromPrivate.js]: `weiAmount` shouldn't be more than account's balance. I've had `0.9995` pdai, but it was displayed as `$1`, so when i've tried to withdraw $1 i've got `Not enough inputs` exception.

Should work:
```es6
await tokenSendV2(
  fromAddress,
  address,
  weiAmount: Math.min(parseFloat(weiAmount), parseFloat(this.state.fromBalance)),
  0,
  xdaiweb3,
  web3,
  metaAccount.privateKey
)
```

**Comment by troggy** [comment]
Just in case, there was a small fix for this functionality in other branch. https://github.com/leapdao/burner-wallet/pull/30/commits/9d097912221008288c2d9d39924b6417f007fb52

I guess because this branch was alrady merged in `sunDai` branch.

**Comment by sunify** [comment]
Thx!

**Comment by TimDaub** [comment]
For now we have #44 to fix for sundai. Should be easily adoptable for master. But for now, it shouldn't be our focus I believe.

**Pull request closed without merging**